File: bench/sieve.go
// Go twin of sieve.ts: a []bool indexed with an int32, so every store is
// bounds-checked exactly as it is in Nish and in the Rust twin (the
// compiler hoists or drops what it can prove).
package main

import "fmt"

const N = 10000000 // bench:n
const PASSES = 20

func sieve(composite []bool, n int32) int32 {
	for i := int32(0); i <= n; i++ {
		composite[i] = false
	}
	for i := int32(2); i*i <= n; i++ {
		if !composite[i] {
			for j := i * i; j <= n; j += i {
				composite[j] = true
			}
		}
	}
	count := int32(0)
	for i := int32(2); i <= n; i++ {
		if !composite[i] {
			count++
		}
	}
	return count
}

func main() {
	composite := make([]bool, N+1)
	total := int32(0)
	for pass := 0; pass < PASSES; pass++ {
		total += sieve(composite, N)
	}
	fmt.Println(total)
}

File: bench/sieve.rs
// Rust twin of sieve.ts: a Vec<bool> indexed with usize, so every store is
// bounds-checked exactly as in Nish (LLVM hoists or drops what it can).
const N: i32 = 10000000; // bench:n
const PASSES: i32 = 20;

fn sieve(composite: &mut [bool], n: i32) -> i32 {
    for i in 0..=n {
        composite[i as usize] = false;
    }
    let mut i: i32 = 2;
    while i * i <= n {
        if !composite[i as usize] {
            let mut j = i * i;
            while j <= n {
                composite[j as usize] = true;
                j += i;
            }
        }
        i += 1;
    }
    let mut count: i32 = 0;
    for k in 2..=n {
        if !composite[k as usize] {
            count += 1;
        }
    }
    count
}

fn main() {
    let mut composite = vec![false; N as usize + 1];
    let mut total: i32 = 0;
    for _ in 0..PASSES {
        total += sieve(&mut composite, N);
    }
    println!("{}", total);
}
